test/CMakeLists.txt

```diff
@@ -0,0 +1,148 @@
+set(CLINT_PRG ${PROJECT_SOURCE_DIR}/src/clint.py)
+set(CLINT_SUPPRESSIONS_DIR ${PROJECT_BINARY_DIR}/errors)
+
+# used for detecting changed files
+if(DEFINED ENV{NVIM_GIT_BASEREF})
+  set(NVIM_GIT_BASEREF "$ENV{NVIM_GIT_BASEREF}")
+else()
+  set(NVIM_GIT_BASEREF "HEAD^1")
+endif()
+
+# this one is used to populate lint-changed-only
+# could also be used as part of a REQUIRED_FIXTURES
+add_test(
+  NAME check_changed_files
+  COMMAND ${CMAKE_COMMAND} -E echo "${changed_files}"
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+)
+
+set_tests_properties(
+  check_changed_files PROPERTIES
+  LABELS lint-changed-only
+)
+
+include(FetchErrorSuppressions)
+FetchErrorSuppressions(
+  TARGET clint-reports
+  DOWNLOAD_DIR ${CLINT_SUPPRESSIONS_DIR}
+)
+
+add_dependencies(nvim clint-reports)
+
+if(CI_BUILD)
+  set(CLINT_OUTPUT_FORMAT gh_action)
+else()
+  set(CLINT_OUTPUT_FORMAT vs7)
+endif()
+
+# add a seprate test for each file for easier invocation
+# e.g. `ctest --test-dir build -R lintuncrustify-src-nvim-window-h`
+foreach(f ${LINT_NVIM_C_FILES})
+  file(RELATIVE_PATH rfile ${PROJECT_SOURCE_DIR} ${f})
+  file(RELATIVE_PATH testname "${PROJECT_SOURCE_DIR}/src/nvim" ${f})
+  string(REGEX REPLACE "[/.]" "-" testname ${testname})
+  set(unc_testname lintuncrustify-${testname})
+  set(clint_testname lintc-${testname})
+  # FIXME(kylo252): this should probably be replaced with a FIXTURES_SETUP
+  if(UNCRUSTIFY_PRG)
+    add_test(
+      NAME ${unc_testname}
+      COMMAND ${UNCRUSTIFY_PRG} -c ${PROJECT_SOURCE_DIR}/src/uncrustify.cfg -q --check ${rfile}
+      WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+    )
+    set_tests_properties(
+      ${unc_testname} PROPERTIES
+      LABELS "lint-all;lintcfull"
+      REQUIRED_FILES ${f}
+    )
+  endif()
+  add_test(
+    NAME ${clint_testname}
+    COMMAND ${CLINT_PRG} --suppress-errors ${CLINT_SUPPRESSIONS_DIR}/${testname}.json --output ${CLINT_OUTPUT_FORMAT} ${rfile}
+    WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+  )
+  set_tests_properties(
+    ${clint_testname} PROPERTIES
+    LABELS "lint-all;lintcfull"
+    REQUIRED_FILES ${f}
+  )
+endforeach()
+
+add_test(
+  NAME lintlua-luacheck
+  COMMAND ${LUACHECK_PRG} -q --no-color runtime/ scripts/ src/ test/
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+)
+
+add_test(
+  NAME lintlua-stylua
+  COMMAND ${STYLUA_PRG}  --color=always --check runtime/
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+)
+
+add_test(
+  NAME lintpy
+  COMMAND ${FLAKE8_PRG} contrib scripts src test
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+)
+
+add_test(
+  NAME lintsh
+  COMMAND ${SHELLCHECK_PRG} scripts/vim-patch.sh
+  WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}
+)
+
+set_tests_properties(
+  lintcommit lintlua-stylua lintlua-luacheck lintpy lintsh PROPERTIES
+  LABELS lint-all
+)
+
+add_custom_target(lintc-changed-only)
+add_custom_command(TARGET lintc-changed-only
+  COMMAND bash -c [==[git diff --name-only --relative=src/nvim HEAD | sed 's@[/.]@-@g' | xargs -n1 ctest --test-dir build -L 'lintc' -R]==]
```

Comment:
This assumes linux only.

Reply:
good point! I guess we need to use cmake to parse the output instead, or make it unix-ish exclusive for now

Reply:
actually, we should rework/(maybe split?) `cmake/Format.cmake`  so we that `get_changed_files` can be used here
https://github.com/neovim/neovim/blob/6da2271a04540f6f3bd6d3cae1895bb1ea1b5011/cmake/Format.cmake#L3-L4

Reply:
actually, we should rework/(maybe split?) `cmake/Format.cmake`  so that `get_changed_files` can be used here
https://github.com/neovim/neovim/blob/6da2271a04540f6f3bd6d3cae1895bb1ea1b5011/cmake/Format.cmake#L3-L4

Reply:
Yeah, makes sense. Maybe we should skip `lintc-changed-only`, and focus on not losing any existing functionality before we start worrying about fancy stuff like this.

Reply:
I figured it'd be a slight regression not to add this, since "touch" files aren't needed/considered anymore.

Overall, I vote for skipping this internal integration for now as well, since running all the lint tests is still extremely fast (takes <10s on my machine)

Reply:
Remove it